agent/agent.go

```diff
@@ -1984,13 +2017,116 @@ func (a *Agent) purgeCheck(checkID types.CheckID) error {
+	Defaults  *persistedServiceConfigResponse
+}
+
+// persistedServiceConfigResponse is a mirror image of
+// *structs.ServiceConfigResponse without QueryMeta fields. These should be
+// kept in sync.
+type persistedServiceConfigResponse struct {
+	ProxyConfig     map[string]interface{}
+	UpstreamConfigs map[string]map[string]interface{}
+	MeshGateway     structs.MeshGatewayConfig `json:",omitempty"`
+}
+
+func (r *persistedServiceConfigResponse) ToResponse() *structs.ServiceConfigResponse {
+	return &structs.ServiceConfigResponse{
+		ProxyConfig:     r.ProxyConfig,
+		UpstreamConfigs: r.UpstreamConfigs,
+		MeshGateway:     r.MeshGateway,
+	}
+}
+
+func (a *Agent) persistServiceConfig(serviceID string, defaults *structs.ServiceConfigResponse) error {
+	// Create the persisted config.
+	wrapped := persistedServiceConfig{
+		ServiceID: serviceID,
+		Defaults: &persistedServiceConfigResponse{
+			ProxyConfig:     defaults.ProxyConfig,
+			UpstreamConfigs: defaults.UpstreamConfigs,
+			MeshGateway:     defaults.MeshGateway,
+		},
+	}
+
+	encoded, err := json.Marshal(wrapped)
+	if err != nil {
+		return err
+	}
+
+	dir := filepath.Join(a.config.DataDir, serviceConfigDir)
+	configPath := filepath.Join(dir, stringHash(serviceID))
+
+	// Create the config dir if it doesn't exist
+	if err := os.MkdirAll(dir, 0700); err != nil {
+		return fmt.Errorf("failed creating service configs dir %q: %s", dir, err)
+	}
+
+	return file.WriteAtomic(configPath, encoded)
+}
+
+func (a *Agent) purgeServiceConfig(serviceID string) error {
+	configPath := filepath.Join(a.config.DataDir, serviceConfigDir, stringHash(serviceID))
+	if _, err := os.Stat(configPath); err == nil {
+		return os.Remove(configPath)
+	}
+	return nil
+}
+
+func (a *Agent) readPersistedServiceConfigs() (map[string]*structs.ServiceConfigResponse, error) {
+	out := make(map[string]*structs.ServiceConfigResponse)
+
+	configDir := filepath.Join(a.config.DataDir, serviceConfigDir)
+	files, err := ioutil.ReadDir(configDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("Failed reading service configs dir %q: %s", configDir, err)
+	}
+
+	for _, fi := range files {
+		// Skip all dirs
+		if fi.IsDir() {
+			continue
+		}
+
+		// Skip all partially written temporary files
+		if strings.HasSuffix(fi.Name(), "tmp") {
+			a.logger.Printf("[WARN] agent: Ignoring temporary service config file %v", fi.Name())
+			continue
+		}
+
+		// Read the contents into a buffer
+		file := filepath.Join(configDir, fi.Name())
+		buf, err := ioutil.ReadFile(file)
+		if err != nil {
+			return nil, fmt.Errorf("failed reading service config file %q: %s", file, err)
+		}
+
+		// Try decoding the service config definition
+		var p persistedServiceConfig
+		if err := json.Unmarshal(buf, &p); err != nil {
+			a.logger.Printf("[ERR] agent: Failed decoding service config file %q: %s", file, err)
+			continue
+		}
+		out[p.ServiceID] = p.Defaults.ToResponse()
+	}
+
+	return out, nil
+}
+
```

Comment:
Thought: I realise this is following the existing pattern, but do you think it would be worth factoring this out into a separate file? This file being big doesn't cause too many practical issues for me so maybe not worth the effort now but I wonder if all the persistence stuff in here would be better as a separate package at some point or at least a separate file in this one?

Reply:
I can move it out into its own file.